// @cloudinary/angular-5.x/src/cloudinary-background-image.directive.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var testing_1 = require("@angular/core/testing");
var platform_browser_1 = require("@angular/platform-browser");
var cloudinary_service_1 = require("./cloudinary.service");
var cloudinary_background_image_directive_1 = require("./cloudinary-background-image.directive");
var cloudinary_transformation_directive_1 = require("./cloudinary-transformation.directive");
var TestComponent = (function () {
    function TestComponent() {
    }
    TestComponent = __decorate([
        core_1.Component({
            template: "\n    <button [clBackgroundImage]=\"image.public_id\" fetch_format=\"auto\" target=\"_blank\">\n        <cl-transformation effect=\"sepia\"></cl-transformation>\n        click me\n    </button>\n    <img [clBackgroundImage]=\"image.public_id\" width=\"100\" crop=\"scale\"/>\n    "
        })
    ], TestComponent);
    return TestComponent;
}());
describe('CloudinaryBackgroundImageDirective', function () {
    var fixture;
    var des; // the three elements w/ the directive
    var localCloudinary = new cloudinary_service_1.Cloudinary(require('cloudinary-core'), { cloud_name: '@@fake_angular2_sdk@@' });
    beforeEach(function () {
        fixture = testing_1.TestBed.configureTestingModule({
            declarations: [cloudinary_background_image_directive_1.CloudinaryBackgroundImageDirective, cloudinary_transformation_directive_1.CloudinaryTransformationDirective, TestComponent],
            providers: [{ provide: cloudinary_service_1.Cloudinary, useValue: localCloudinary }]
        }).createComponent(TestComponent);
        var comp = fixture.componentInstance;
        // pretend that it was wired to something that supplied an image
        comp['image'] = {
            public_id: 'some_image_id'
        };
        fixture.detectChanges(); // initial binding
        // all elements with an attached CloudinaryImageSourceDirective
        des = fixture.debugElement.queryAll(platform_browser_1.By.directive(cloudinary_background_image_directive_1.CloudinaryBackgroundImageDirective));
    });
    it('should have 1 element', function () {
        expect(des.length).toBe(2);
    });
    it('updates the background image with transformations', function () {
        var button = des[0].nativeElement;
        expect(button.style.backgroundImage).toEqual(jasmine.stringMatching(/image\/upload\/e_sepia\/f_auto\/some_image_id/));
        expect(button.textContent).toEqual(jasmine.stringMatching(/click me/));
    });
    it('updates the background image without transformations', function () {
        var div = des[1].nativeElement;
        expect(div.style.backgroundImage).toEqual(jasmine.stringMatching(/c_scale,w_100\/some_image_id/));
    });
});
