// @cloudinary/angular-5.x/src/cloudinary-image.component.spec.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var testing_1 = require("@angular/core/testing");
var platform_browser_1 = require("@angular/platform-browser");
var cloudinary_service_1 = require("./cloudinary.service");
var cloudinary_image_component_1 = require("./cloudinary-image.component");
var cloudinary_transformation_directive_1 = require("./cloudinary-transformation.directive");
describe('CloudinaryImage', function () {
    var localCloudinary = new cloudinary_service_1.Cloudinary(require('cloudinary-core'), { cloud_name: '@@fake_angular2_sdk@@' });
    beforeEach(function () {
        spyOn(localCloudinary, 'toCloudinaryAttributes').and.callThrough();
        spyOn(localCloudinary, 'url').and.callThrough();
        spyOn(localCloudinary, 'responsive').and.callThrough();
    });
    describe('responsive images without nested transformations', function () {
        var TestComponent = (function () {
            function TestComponent() {
            }
            TestComponent = __decorate([
                core_1.Component({
                    template: "<cl-image responsive id=\"image1\" width=\"300\" crop=\"scale\" effect=\"blackwhite\" public-id=\"responsive_sample.jpg\"></cl-image>"
                })
            ], TestComponent);
            return TestComponent;
        }());
        var fixture;
        var des; // the elements w/ the directive
        beforeEach(function () {
            fixture = testing_1.TestBed.configureTestingModule({
                declarations: [cloudinary_image_component_1.CloudinaryImage, TestComponent],
                providers: [{ provide: cloudinary_service_1.Cloudinary, useValue: localCloudinary }]
            }).createComponent(TestComponent);
            fixture.detectChanges(); // initial binding
            expect(localCloudinary.responsive).toHaveBeenCalled();
            // all elements with an attached CloudinaryImage
            des = fixture.debugElement.query(platform_browser_1.By.directive(cloudinary_image_component_1.CloudinaryImage));
        });
        it('creates an img element which encodes the directive attributes to the URL', function () {
            var img = des.children[0].nativeElement;
            // http://res.cloudinary.com/@@fake_angular2_sdk@@/image/upload/c_scale,e_blackwhite,w_300/responsive_sample.jpg
            expect(img.attributes.getNamedItem('src').value).toEqual(jasmine.stringMatching(/c_scale,e_blackwhite,w_300\/responsive_sample.jpg/));
            expect(img.attributes.getNamedItem('data-src').value).toEqual(jasmine.stringMatching(/c_scale,e_blackwhite,w_300\/responsive_sample.jpg/));
        });
    });
    describe('responsive images with nested transformations', function () {
        var TestComponent = (function () {
            function TestComponent() {
            }
            TestComponent = __decorate([
                core_1.Component({
                    template: "<cl-image responsive id=\"image1\" public-id=\"responsive_sample.jpg\">\n            <cl-transformation width=\"300\" crop=\"scale\" overlay=\"text:roboto_25_bold:SDK\"></cl-transformation>\n            <cl-transformation effect=\"art:hokusai\" gravity=\"auto\"></cl-transformation>\n            <cl-transformation fetch-format=\"auto\"></cl-transformation>\n            </cl-image>"
                })
            ], TestComponent);
            return TestComponent;
        }());
        var fixture;
        var des; // the elements w/ the directive
        beforeEach(function () {
            fixture = testing_1.TestBed.configureTestingModule({
                declarations: [cloudinary_transformation_directive_1.CloudinaryTransformationDirective, cloudinary_image_component_1.CloudinaryImage, TestComponent],
                providers: [{ provide: cloudinary_service_1.Cloudinary, useValue: localCloudinary }]
            }).createComponent(TestComponent);
            fixture.detectChanges(); // initial binding
            // all elements with an attached CloudinaryImage
            des = fixture.debugElement.query(platform_browser_1.By.directive(cloudinary_image_component_1.CloudinaryImage));
        });
        it('creates an img element which encodes the directive attributes to the URL', function () {
            var img = des.children[0].nativeElement;
            expect(img.src).toEqual(jasmine.stringMatching(/c_scale,l_text:roboto_25_bold:SDK,w_300\/e_art:hokusai,g_auto\/f_auto\/responsive_sample.jpg/));
            expect(img.attributes.getNamedItem('data-src').value).toEqual(jasmine.stringMatching(/c_scale,l_text:roboto_25_bold:SDK,w_300\/e_art:hokusai,g_auto\/f_auto\/responsive_sample.jpg/));
        });
    });
    describe('images with overlay/underlay', function () {
        var TestComponent = (function () {
            function TestComponent() {
            }
            TestComponent = __decorate([
                core_1.Component({
                    template: "\n        <cl-image responsive id=\"image1\" public-id=\"responsive_sample.jpg\">\n          <cl-transformation overlay=\"fetch:http://cloudinary.com/images/old_logo.png\"></cl-transformation>\n          <cl-transformation underlay=\"fetch:http://cloudinary.com/images/old_logo.png\"></cl-transformation>\n        </cl-image>\n        <cl-image responsive id=\"image2\" public-id=\"responsive_sample.jpg\">\n          <cl-transformation overlay=\"fetch:https://upload.wikimedia.org/wikipedia/commons/2/2b/\uACE0\uCC3D\uAC2F\uBC8C.jpg\"></cl-transformation>\n          <cl-transformation underlay=\"fetch:https://upload.wikimedia.org/wikipedia/commons/2/2b/\uACE0\uCC3D\uAC2F\uBC8C.jpg\"></cl-transformation>\n        </cl-image>\n      "
                })
            ], TestComponent);
            return TestComponent;
        }());
        var fixture;
        var des;
        beforeEach(function () {
            fixture = testing_1.TestBed.configureTestingModule({
                declarations: [cloudinary_transformation_directive_1.CloudinaryTransformationDirective, cloudinary_image_component_1.CloudinaryImage, TestComponent],
                providers: [{ provide: cloudinary_service_1.Cloudinary, useValue: localCloudinary }]
            }).createComponent(TestComponent);
            fixture.detectChanges(); // initial binding
            // all elements with an attached CloudinaryImage
            des = fixture.debugElement.queryAll(platform_browser_1.By.directive(cloudinary_image_component_1.CloudinaryImage));
        });
        it('should serialize a fetch URL', function () {
            var img = des[0].children[0].nativeElement;
            expect(img.src).toEqual(jasmine.stringMatching(/l_fetch:aHR0cDovL2Nsb3VkaW5hcnkuY29tL2ltYWdlcy9vbGRfbG9nby5wbmc=\/u_fetch:aHR0cDovL2Nsb3VkaW5hcnkuY29tL2ltYWdlcy9vbGRfbG9nby5wbmc=\/responsive_sample.jpg/));
            expect(img.attributes.getNamedItem('data-src').value).toEqual(jasmine.stringMatching(/l_fetch:aHR0cDovL2Nsb3VkaW5hcnkuY29tL2ltYWdlcy9vbGRfbG9nby5wbmc=\/u_fetch:aHR0cDovL2Nsb3VkaW5hcnkuY29tL2ltYWdlcy9vbGRfbG9nby5wbmc=\/responsive_sample.jpg/));
        });
        it('should support unicode URLs', function () {
            var img = des[1].children[0].nativeElement;
            expect(img.src).toEqual(jasmine.stringMatching(new RegExp('l_fetch:aHR0cHM6Ly91cGxvYWQud2lraW1lZGlhLm9yZy93aWtpcGVkaWEvY29tbW9ucy8yLzJiLyVFQSVCMyVB' +
                'MCVFQyVCMCVCRCVFQSVCMCVBRiVFQiVCMiU4Qy5qcGc=/u_fetch:aHR0cHM6Ly91cGxvYWQud2lraW1lZGlhLm9yZy93aWtpcGVkaWEv' +
                'Y29tbW9ucy8yLzJiLyVFQSVCMyVBMCVFQyVCMCVCRCVFQSVCMCVBRiVFQiVCMiU4Qy5qcGc=/responsive_sample.jpg')));
            expect(img.attributes.getNamedItem('data-src').value).toEqual(jasmine.stringMatching(new RegExp('l_fetch:aHR0cHM6Ly91cGxvYWQud2lraW1lZGlhLm9yZy93aWtpcGVkaWEvY29tbW9ucy8yLzJiLyVFQSVCMyVB' +
                'MCVFQyVCMCVCRCVFQSVCMCVBRiVFQiVCMiU4Qy5qcGc=/u_fetch:aHR0cHM6Ly91cGxvYWQud2lraW1lZGlhLm9yZy93aWtpcGVkaWEv' +
                'Y29tbW9ucy8yLzJiLyVFQSVCMyVBMCVFQyVCMCVCRCVFQSVCMCVBRiVFQiVCMiU4Qy5qcGc=/responsive_sample.jpg')));
        });
    });
    describe('transformation attributes: quality', function () {
        var TestComponent = (function () {
            function TestComponent() {
            }
            TestComponent = __decorate([
                core_1.Component({
                    template: "\n        <cl-image responsive id=\"image1\" public-id=\"responsive_sample.jpg\">\n          <cl-transformation quality=\"0.4\"></cl-transformation>\n        </cl-image>\n        <cl-image responsive id=\"image2\" public-id=\"responsive_sample.jpg\">\n          <cl-transformation quality=\"auto\"></cl-transformation>\n        </cl-image>\n        <cl-image responsive id=\"image3\" public-id=\"responsive_sample.jpg\">\n          <cl-transformation quality=\"auto:good\"></cl-transformation>\n        </cl-image>\n      "
                })
            ], TestComponent);
            return TestComponent;
        }());
        var fixture;
        var des; // the elements w/ the directive
        beforeEach(function () {
            fixture = testing_1.TestBed.configureTestingModule({
                declarations: [cloudinary_transformation_directive_1.CloudinaryTransformationDirective, cloudinary_image_component_1.CloudinaryImage, TestComponent],
                providers: [{ provide: cloudinary_service_1.Cloudinary, useValue: localCloudinary }]
            }).createComponent(TestComponent);
            fixture.detectChanges(); // initial binding
            // all elements with an attached CloudinaryImage
            des = fixture.debugElement.queryAll(platform_browser_1.By.directive(cloudinary_image_component_1.CloudinaryImage));
        });
        it('creates an img element which encodes the quality parameter to URL', function () {
            var testResults = ['q_0.4', 'q_auto', 'q_auto:good'];
            testResults.forEach(function (result, index) {
                var img = des[index].children[0].nativeElement;
                expect(img.src).toEqual(jasmine.stringMatching(new RegExp("/" + result + "/responsive_sample.jpg")));
            });
        });
    });
    describe('missing public-id', function () {
        var TestComponent = (function () {
            function TestComponent() {
            }
            TestComponent = __decorate([
                core_1.Component({
                    template: '<cl-image responsive id="image1"></cl-image>'
                })
            ], TestComponent);
            return TestComponent;
        }());
        it('throws if the directive is missing a public-id attribute', function () {
            var fixture = testing_1.TestBed.configureTestingModule({
                declarations: [cloudinary_transformation_directive_1.CloudinaryTransformationDirective, cloudinary_image_component_1.CloudinaryImage, TestComponent],
                providers: [{ provide: cloudinary_service_1.Cloudinary, useValue: localCloudinary }]
            }).createComponent(TestComponent);
            expect(function () { fixture.detectChanges(); }).toThrowError(/You must set the public id of the image to load/i);
        });
    });
    describe('non-responsive images with nested transformations', function () {
        var TestComponent = (function () {
            function TestComponent() {
            }
            TestComponent = __decorate([
                core_1.Component({
                    template: "<cl-image id=\"image1\" public-id=\"responsive_sample.jpg\">\n            <cl-transformation width=\"300\" crop=\"scale\" overlay=\"text:roboto_35_bold:SDK\"></cl-transformation>\n            <cl-transformation effect=\"art:hokusai\"></cl-transformation>\n            <cl-transformation fetch-format=\"auto\"></cl-transformation>\n            <cl-transformation if=\"initialWidth > 400\" effect=\"grayscale\"></cl-transformation>\n            <cl-transformation if=\"initialHeight < 200\" effect=\"blur\"></cl-transformation>\n            </cl-image>"
                })
            ], TestComponent);
            return TestComponent;
        }());
        var fixture;
        var des; // the elements w/ the directive
        beforeEach(function () {
            fixture = testing_1.TestBed.configureTestingModule({
                declarations: [cloudinary_transformation_directive_1.CloudinaryTransformationDirective, cloudinary_image_component_1.CloudinaryImage, TestComponent],
                providers: [{ provide: cloudinary_service_1.Cloudinary, useValue: localCloudinary }]
            }).createComponent(TestComponent);
            fixture.detectChanges(); // initial binding
            // all elements with an attached CloudinaryImage
            des = fixture.debugElement.query(platform_browser_1.By.directive(cloudinary_image_component_1.CloudinaryImage));
        });
        it('creates an img element which encodes the directive attributes to the URL', function () {
            var img = des.children[0].nativeElement;
            expect(img.src).toEqual(jasmine.stringMatching(/c_scale,l_text:roboto_35_bold:SDK,w_300\/e_art:hokusai\/f_auto\/if_iw_gt_400,e_grayscale\/if_ih_lt_200,e_blur\/responsive_sample.jpg/));
            expect(img.attributes.getNamedItem('data-src')).toBeNull();
        });
        it('updates the underlying img dynamically by updating attributes', function (done) {
            // Couldn't get this to work with Angular's async or fakesync
            // Add another mutation observer on the node to be able to
            // verify the change
            var observer = new MutationObserver(function () {
                var img = des.children[0].nativeElement;
                expect(img.src).toEqual(jasmine.stringMatching(/c_scale,l_text:roboto_35_bold:SDK,w_300\/e_art:hokusai\/f_auto\/if_iw_gt_400,e_grayscale\/if_ih_lt_200,e_blur\/o_50\/responsive_sample.jpg/));
                observer.disconnect();
                done();
            });
            // Observe changes to attributes or child transformations to re-render the image
            var config = { attributes: true, childList: true };
            // pass in the target node, as well as the observer options
            observer.observe(des.nativeElement, config);
            des.nativeElement.setAttribute('opacity', '50');
        });
    });
    describe('Sample code presented in README', function () {
        var TestComponent = (function () {
            function TestComponent() {
            }
            TestComponent = __decorate([
                core_1.Component({
                    template: "\n            <cl-image public-id=\"readme\" class=\"thumbnail inline\" angle=\"20\" format=\"jpg\">\n                <cl-transformation height=\"150\" width=\"150\" crop=\"fill\" effect=\"sepia\" radius=\"20\"></cl-transformation>\n                <cl-transformation overlay=\"text:arial_60:readme\" gravity=\"north\" y=\"20\"></cl-transformation>\n            </cl-image>\n            "
                })
            ], TestComponent);
            return TestComponent;
        }());
        var fixture;
        var des; // the elements w/ the directive
        beforeEach(function () {
            fixture = testing_1.TestBed.configureTestingModule({
                declarations: [cloudinary_transformation_directive_1.CloudinaryTransformationDirective, cloudinary_image_component_1.CloudinaryImage, TestComponent],
                providers: [{ provide: cloudinary_service_1.Cloudinary, useValue: localCloudinary }]
            }).createComponent(TestComponent);
            fixture.detectChanges(); // initial binding
            // Our element under test, which is attached to CloudinaryImage
            des = fixture.debugElement.query(platform_browser_1.By.directive(cloudinary_image_component_1.CloudinaryImage));
        });
        it('creates an img element which encodes the directive attributes to the URL', function () {
            var img = des.children[0].nativeElement;
            expect(img.src).toEqual(jasmine.stringMatching(/c_fill,e_sepia,h_150,r_20,w_150\/g_north,l_text:arial_60:readme,y_20\/a_20\/readme.jpg/));
            expect(img.attributes.getNamedItem('data-src')).toBeNull();
        });
    });
    describe('Bound public-id', function () {
        var TestComponent = (function () {
            function TestComponent() {
                this.publicId = 'sample';
            }
            TestComponent = __decorate([
                core_1.Component({
                    template: "<cl-image id=\"image1\" [public-id]=\"publicId\"></cl-image>"
                })
            ], TestComponent);
            return TestComponent;
        }());
        var fixture;
        var des; // the elements w/ the directive
        beforeEach(function () {
            fixture = testing_1.TestBed.configureTestingModule({
                declarations: [cloudinary_transformation_directive_1.CloudinaryTransformationDirective, cloudinary_image_component_1.CloudinaryImage, TestComponent],
                providers: [{ provide: cloudinary_service_1.Cloudinary, useValue: localCloudinary }]
            }).createComponent(TestComponent);
            fixture.detectChanges(); // initial binding
            // all elements with an attached CloudinaryImage
            des = fixture.debugElement.query(platform_browser_1.By.directive(cloudinary_image_component_1.CloudinaryImage));
        });
        it('creates an img element with a bound public-id', function () {
            var img = des.children[0].nativeElement;
            expect(img.src).toEqual(jasmine.stringMatching(/image\/upload\/sample/));
            // Update data-bound publicId
            fixture.componentInstance.publicId = 'updatedId';
            fixture.detectChanges();
            // Verify that the img src has updated
            expect(img.src).toEqual(jasmine.stringMatching(/image\/upload\/updatedId/));
        });
    });
    describe('event emitters', function () {
        var onImageLoad;
        var onImageError;
        var des; // the elements w/ the directive
        var TestComponent = (function () {
            function TestComponent() {
                this.publicId = 'sample';
            }
            TestComponent = __decorate([
                core_1.Component({
                    template: "<cl-image id=\"image1\" [public-id]=\"publicId\"></cl-image>"
                })
            ], TestComponent);
            return TestComponent;
        }());
        var fixture;
        beforeEach(function () {
            fixture = testing_1.TestBed.configureTestingModule({
                declarations: [cloudinary_transformation_directive_1.CloudinaryTransformationDirective, cloudinary_image_component_1.CloudinaryImage, TestComponent],
                providers: [{ provide: cloudinary_service_1.Cloudinary, useValue: localCloudinary }]
            }).createComponent(TestComponent);
            fixture.detectChanges(); // initial binding
            // all elements with an attached CloudinaryImage
            des = fixture.debugElement.query(platform_browser_1.By.directive(cloudinary_image_component_1.CloudinaryImage));
            onImageLoad = jasmine.createSpy('onImageLoad');
            onImageError = jasmine.createSpy('onImageError');
            des.componentInstance.onLoad.subscribe(onImageLoad);
            des.componentInstance.onError.subscribe(onImageError);
        });
        it('calls the onLoad callback when image loads successfully', function () {
            // Simulate the load event
            var img = des.children[0].nativeElement;
            img.dispatchEvent(new Event('load'));
            expect(onImageLoad).toHaveBeenCalled();
            expect(onImageError).not.toHaveBeenCalled();
        });
        it('calls the onError callback when image fails to load', function () {
            // Simulate the error event
            var img = des.children[0].nativeElement;
            img.dispatchEvent(new CustomEvent('error', { detail: 'Eitan was here' }));
            expect(onImageLoad).not.toHaveBeenCalled();
            expect(onImageError).toHaveBeenCalled();
        });
    });
    describe('responsive images with nested transformations using the cld-responsive attribute', function () {
        var TestComponent = (function () {
            function TestComponent() {
            }
            TestComponent = __decorate([
                core_1.Component({
                    template: "<cl-image cld-responsive id=\"image1\" public-id=\"responsive_sample.jpg\">\n             <cl-transformation width=\"300\" crop=\"scale\" overlay=\"text:roboto_25_bold:SDK\"></cl-transformation>\n             <cl-transformation effect=\"art:hokusai\"></cl-transformation>\n             <cl-transformation fetch-format=\"auto\"></cl-transformation>\n             </cl-image>"
                })
            ], TestComponent);
            return TestComponent;
        }());
        var fixture;
        var des; // the elements w/ the directive
        beforeEach(function () {
            fixture = testing_1.TestBed.configureTestingModule({
                declarations: [cloudinary_transformation_directive_1.CloudinaryTransformationDirective, cloudinary_image_component_1.CloudinaryImage, TestComponent],
                providers: [{ provide: cloudinary_service_1.Cloudinary, useValue: localCloudinary }]
            }).createComponent(TestComponent);
            fixture.detectChanges(); // initial binding
            expect(localCloudinary.responsive).toHaveBeenCalled();
            // all elements with an attached CloudinaryImage
            des = fixture.debugElement.query(platform_browser_1.By.directive(cloudinary_image_component_1.CloudinaryImage));
        });
        it('creates an img element which encodes the directive attributes to the URL', function () {
            var img = des.children[0].nativeElement;
            expect(img.src).toEqual(jasmine.stringMatching(/c_scale,l_text:roboto_25_bold:SDK,w_300\/e_art:hokusai\/f_auto\/responsive_sample.jpg/));
            expect(img.attributes.getNamedItem('data-src').value).toEqual(jasmine.stringMatching(/c_scale,l_text:roboto_25_bold:SDK,w_300\/e_art:hokusai\/f_auto\/responsive_sample.jpg/));
        });
    });
});
